engine/src/graphics/device.cpp

```diff
@@ -59,13 +69,103 @@ namespace gen
 		VULKAN_HPP_DEFAULT_DISPATCHER.init(*m_instance);
 #endif
 	}
-	void GraphicsDevice::createSurface(Window const &window)
+  
+	void GraphicsDevice::createSurface(const Window & window)
 	{
 		m_surface = vk::util::createWindowSurface(m_instance.get(), window);
 		gen::logger::debug("vulkan", "Created surface");
 	}
+	void GraphicsDevice::pickPhysicalDevice()
+	{
+		// TODO: Add a weighting system to pick the best physical device.
+		// TODO: Add a way to specify the physical device to use if not using the default.
+		auto availablePhysicalDevices = m_instance->enumeratePhysicalDevices();
+		if (availablePhysicalDevices.empty())
+        {
+			gen::logger::error("vulkan", "Failed to find GPUs with Vulkan support!");
+            throw std::runtime_error("Failed to find GPUs with Vulkan support!");
+        }
+
+		m_physicalDevice = availablePhysicalDevices.front();
```

Comment:
Instead of picking the first one, which may be integrated, better to try to select a discrete GPU if present, otherwise fallback to `.front()`. Beyond that there isn't really much need to weigh the best device, especially if users can customize which GPU to use.